### PR TITLE
[GEP-26] Enable validation of `seed.spec.backup.credentialsRef` and `backupBucket.spec.credentialsRef`

### PR DESCRIPTION
**Pull request by vpnachev**
**How to categorize this PR?**
/area security ipcei
/kind enhancement
/platform gcp

**What this PR does / why we need it**:
Enable validation of `seed.spec.backup.credentialsRef` and `backupBucket.spec.credentialsRef`

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9586

**Special notes for your reviewer**:
cc @dimityrmirchev 

**Release note**:

```feature operator
Admission controller now allows `seed.spec.backup.credentialsRef` and `backupBucket.spec.credentialsRef` to use credentials of either `v1.Secret` or `security.gardener.cloud/v1alpha1.WorkloadIdentity`
```

### CODE REVIEW

File: pkg/apis/gcp/validation/backupbucket.go [hunk 1]
```diff
@@ -31,3 +35,56 @@ func ValidateBackupBucketConfig(config *apisgcp.BackupBucketConfig, fldPath *fie
...
+	var (
+		allErrs          = field.ErrorList{}
+		immutabilityPath = fldPath.Child("immutability")
+	)
+
+	if oldConfig.Immutability == nil || !oldConfig.Immutability.Locked {
+		return allErrs
+	}
+
+	if newConfig == nil || newConfig.Immutability == nil || *newConfig.Immutability == (apisgcp.ImmutableConfig{}) {
```
Comment (dimityrmirchev) [inline pkg/apis/gcp/validation/backupbucket.go:50]: Do we need the brackets?
```suggestion
	if newConfig == nil || newConfig.Immutability == nil || *newConfig.Immutability == apisgcp.ImmutableConfig{} {
```

File: pkg/apis/gcp/validation/backupbucket_test.go [hunk 2]
```diff
@@ -9,77 +9,210 @@ import (
...
-				},
-			}, true, "must be a positive duration greater than 24h"),
-	)
+			errs := ValidateBackupBucketCredentialsRef(credsRef, fldPath)
+			Expect(errs).To(BeEmpty())
+		})
+	})
 })
+
+func generateBackupBucketConfig(retentionType string, retentionPeriod *time.Duration, locked bool, isImmutableConfigured bool) apisgcp.BackupBucketConfig {
```
Comment (dimityrmirchev) [inline pkg/apis/gcp/validation/backupbucket_test.go:201]: I would call this either `isImmutabilityConfigured` or `isImmutable`. I am not sure which fits better.
```suggestion
func generateBackupBucketConfig(retentionType string, retentionPeriod *time.Duration, locked bool, isImmutabilityConfigured bool) apisgcp.BackupBucketConfig {
```

File: pkg/apis/gcp/validation/backupbucket.go (hunk 1 above)
Comment (vpnachev) [reply pkg/apis/gcp/validation/backupbucket.go:50]: Looks like the linter is not happy without the brackets
```bash
$ make verify
...
> Check
Executing golangci-lint
pkg/apis/gcp/validation/backupbucket.go:1: : # github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/validation [github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/validation.test]
pkg/apis/gcp/validation/backupbucket.go:50:111: syntax error: unexpected { at end of statement
pkg/apis/gcp/validation/backupbucket.go:54:2: syntax error: non-declaration statement outside function body (typecheck)
// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
pkg/apis/gcp/validation/backupbucket.go:50:111: expected ';', found '{' (typecheck)
        if newConfig == nil || newConfig.Immutability == nil || *newConfig.Immutability == apisgcp.ImmutableConfig{} {
                                                                                                                     ^
pkg/apis/gcp/validation/backupbucket.go:54:2: expected declaration, found 'if' (typecheck)
        if !newConfig.Immutability.Locked {
        ^
pkg/apis/gcp/validation/backupbucket.go:85:2: expected declaration, found 'if' (typecheck)
        if !allowedGVKs.Has(credentialsRef.GroupVersionKind()) {
        ^
...
```

File: pkg/admission/validator/backupbucket_test.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+				newBackupBucket := backupBucket.DeepCopy()
+				newBackupBucket.Spec.ProviderConfig = &runtime.RawExtension{
+					Raw: []byte(`{"apiVersion": "gcp.provider.extensions.gardener.cloud/v1alpha1", "kind": "invalid"}`),
+				}
+
+				err := backupBucketValidator.Validate(ctx, newBackupBucket, backupBucket)
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(MatchError(ContainSubstring(`failed to decode new provider config: no kind "invalid" is registered for version "gcp.provider.extensions.gardener.cloud/v1alpha1" in scheme`)))
+			})
+
```
Comment (hebelsan) [inline pkg/admission/validator/backupbucket_test.go:116]: Could you add a test where providerConfig is initially set, but then set to nil?

**Review by hebelsan — COMMENTED** [review]
Thanks for the PR!
LGTM, just a small nit.

File: pkg/admission/validator/backupbucket_test.go (hunk 3 above)
Comment (hebelsan) [reply pkg/admission/validator/backupbucket_test.go:116]: It would unit test `DecodeBackupBucketConfig` with a `nil` input in combination with validating the `newSeed` update.
I think it makes sense to cover this scenario to guard against regressions from future changes.

Comment (vpnachev) [reply pkg/admission/validator/backupbucket_test.go:116]: A test is added with 8c508ac5740d4, but to be honest, I think internals of `DecodeBackupBucketConfig` should and are tested in the `pkg/admission/decode_test.go`.

**Review by hebelsan — APPROVED** [review]
/lgtm

**Review by dimityrmirchev — APPROVED** [review]
/lgtm
